### PR TITLE
Fix base class should be explicitly initialized in copy constructor

### PR DESCRIPTION
**Pull request by SunBlack**

### CODE REVIEW

**Comment by taketwo** [comment]
Can we close this in favor of #3075 or is there anything that was not addressed there?

**Comment by SunBlack** [comment]
As I touched 18 files and we have only 12 conflicts, there must be something open. I check it later.

**Comment by SunBlack** [comment]
It is mostly a cleanup PR now, even it contains still some warning fixes.

Btw suggestion for the style guide:
- In case copy & assignment constructor are not needed, the assignment constructor should be deleted (personal preference by me, but I think we should have a guideline there)
- Deleted functions should be public, because normal you are looking in public member for special members. So it increases readability.

**Comment by SergioRAgostinho** [comment]
>     * In case copy & assignment constructor are not needed, the assignment constructor should be deleted (personal preference by me, but I think we should have a guideline there)

Can you rephrase this sentence revising the terms you used. Just to avoid incorrect interpretation
```cpp
Class(const Class& other) // copy constructor

Class& operator=(const Class& other) // copy assignment
```
Something is either an assignment operator or a constructor but never an **assignment constructor**.

**Comment by SunBlack** [comment]
Sorry, I meant I prefer deleting copy constructor over deleting copy assignment.

**Comment by SergioRAgostinho** [comment]
> Sorry, I meant I prefer deleting copy constructor over deleting copy assignment.

The general case is the following: if one needs a user defined copy constructor, it also needs a user defined copy assignment. I've never faced a situation in which you need one and not the other. So my suggestion is: either both or none should be defined.



>   Deleted functions should be public, because normal you are looking in public member for special members. So it increases readability.

Valid point. On this one I personally prefer to not have implicitly deleted methods on the headers. Less code to maintain.

**Comment by SunBlack** [comment]
Had now finally time to fix last compile issues. So it's ready to review :).
